### PR TITLE
Upgrade to latest Level modules

### PR DESCRIPTION
**Pull request by vweevers**
Semver-major. Goes hand-in-hand with a `level-party` PR which I'll open shortly. Achieves the following:

- Upgrade to latest Level modules
  - Mainly to support node 10, 12 and future
  - Also includes latest `memdown`, which now internally stores keys+values as Buffers, making it more reliable as a `leveldown` replacement
- Support an input `db` that isn't `level` (may or may not have `levelup`, `encoding-down`, `deferred-leveldown` layers, and may have additional layers)
- Expose all methods that `level-party` needs on the `multileveldown` client, so that `level-party` doesn't have to reach down (e.g. `db.db`)
- Support `subleveldown` on `level-party`
- Support `subleveldown` on `multileveldown` client (i.e. for client-side sublevels)
- Support `multileveldown` server on `subleveldown` (i.e. to expose only a sublevel to client)

### CODE REVIEW

File: leveldown.js [hunk 1]
```diff
@@ -21,9 +23,9 @@ var DECODERS = [
 
 module.exports = Multilevel
 
-function Multilevel (path, opts) {
-  if (!(this instanceof Multilevel)) return new Multilevel(path, opts)
-  abstract.AbstractLevelDOWN.call(this, path)
+function Multilevel (opts) {
+  if (!(this instanceof Multilevel)) return new Multilevel(opts)
+  abstract.AbstractLevelDOWN.call(this)
```
Comment (vweevers) [inline leveldown.js]: `abstract-leveldown` no longer needs a location (aka `path` here).

File: leveldown.js [hunk 2]
```diff
@@ -38,6 +40,8 @@ function Multilevel (path, opts) {
 
 util.inherits(Multilevel, abstract.AbstractLevelDOWN)
 
+Multilevel.prototype.type = 'multileveldown'
```
Comment (vweevers) [inline leveldown.js]: This is for compatibility with [`reachdown`](https://github.com/vweevers/reachdown), which I also intend to start using in `subleveldown`, `level-test` and more.

File: leveldown.js [hunk 3]
```diff
@@ -112,12 +116,12 @@ Multilevel.prototype.createRpcStream = function (opts, proxy) {
...
   function oncallback (res) {
     var req = self._requests.remove(res.id)
-    if (req) req.callback(decodeError(res.error), decodeValue(res.value, req.valueEncoding))
+    if (req) req.callback(decodeError(res.error), decodeValue(res.value, req.valueAsBuffer))
   }
 }
 
 Multilevel.prototype.forward = function (down) {
-  this._db = down
+  this._db = reachdown(down, matchdown, false)
```
Comment (vweevers) [inline leveldown.js]: This peels off `levelup`,` deferred-leveldown` and `encoding-down`.

File: leveldown.js [hunk 4]
```diff
@@ -145,14 +149,22 @@ Multilevel.prototype._clearRequests = function (closing) {
   }
 }
 
+Multilevel.prototype._serializeKey = function (key) {
+  return Buffer.isBuffer(key) ? key : Buffer.from(String(key))
+}
+
+Multilevel.prototype._serializeValue = function (value) {
+  return Buffer.isBuffer(value) ? value : Buffer.from(String(value))
+}
```
Comment (vweevers) [inline leveldown.js]: This ensures we're only working with buffers internally.

File: leveldown.js [hunk 5]
```diff
@@ -145,14 +149,22 @@ Multilevel.prototype._clearRequests = function (closing) {
...
+
 Multilevel.prototype._get = function (key, opts, cb) {
   if (this._db) return this._db._get(key, opts, cb)
 
   var req = {
     tag: 0,
     id: 0,
     key: key,
-    valueEncoding: opts.valueEncoding || (opts.asBuffer === false ? 'utf-8' : 'binary'),
+    valueAsBuffer: opts.asBuffer,
```
Comment (vweevers) [inline leveldown.js]: We no longer have access to `key/ValueEncoding` options here, but we don't need it. We only have to care about returning a buffer or string (which at a higher level, is dictated by `encoding-down` and the user's choice of encoding).

File: server.js [hunk 6]
```diff
@@ -29,7 +32,7 @@ module.exports = function (db, opts) {
   return stream
 
   function ready () {
-    var down = db.db
+    var down = reachdown(db, matchdown, false)
```
Comment (vweevers) [inline server.js]: This peels off `levelup`, `deferred-leveldown` and `encoding-down`.

File: server.js [hunk 7]
```diff
@@ -136,15 +140,7 @@ function Iterator (down, req, encode) {
...
-
-  if (req.options) {
-    if (req.options.gt === null) req.options.gt = undefined
-    if (req.options.gte === null) req.options.gte = undefined
-    if (req.options.lt === null) req.options.lt = undefined
-    if (req.options.lte === null) req.options.lte = undefined
-  }
-
-  this._iterator = down.iterator(req.options)
+  this._iterator = down.iterator(cleanRangeOptions(req.options))
```
Comment (vweevers) [inline server.js]: Because `null` and `undefined` are valid range options in latest `abstract-leveldown`, we must strip them from the options object.

**Comment by julien-f** [comment]
@vweevers Any chance of seeing this and `level-party` merged?

Do you think we should do a proper fork?

**Comment by vweevers** [comment]
We can move these modules to Level if @mafintosh is cool with that.

**Comment by julien-f** [comment]
Would be really nice, I'm currently stuck on Node 8 because I'm using `level-party`.

**Comment by mafintosh** [comment]
I’m very cool with that!

**Comment by vweevers** [comment]
Thanks @mafintosh! Do we need @substack for moving `level-party`?

**Pull request merged**